### PR TITLE
improve(TxUtils): Should try to resubmit txn when encountering certain errors

### PR DESCRIPTION
**Pull request by nicholaspai**
For  nonce collisions and gas price estimation errors, bots should try to resubmit

### CODE REVIEW

File: src/utils/TransactionUtils.ts [hunk 1]
```diff
@@ -47,14 +47,13 @@ export async function runTransaction(
...
-      // If error is `insufficient funds for intrinsic transaction cost` the gas price is probably spiking, just
-      // resubmit to get the new gas price.
-      return await runTransaction(logger, contract, method, args, value, gasLimit, nonce);
+    if (
+      error?.code === "NONCE_EXPIRED" ||
+      error?.code === "INSUFFICIENT_FUNDS" ||
+      error?.message.includes("intrinsic gas too low")
+    ) {
+      // If error is due to a nonce collision or gas underpricement then re-submit to fetch latest params.
+      return await runTransaction(logger, contract, method, args, value, gasLimit, null);
```
Comment (pxrl) [inline src/utils/TransactionUtils.ts:56]: In general I think we should protect against the potential for endless recursion here. It seems like there's a scenario where a txn could pass the initial simulation, but then end up in a recursive INSUFFICIENT_FUNDS loop due to a gas price spike coinciding with insufficient funds to pay the new gas price.

One possible way of mitigating that would be to modify the input arguments to runTransaction(), such that it accepts a counter that is decremented on subsequent calls into runTransaction(), or a boolean flag that determines whether it can recursively call down to runTransaction().

```
      return await runTransaction(logger, contract, method, args, value, gasLimit, null, --permitRetry);
```

Comment (nicholaspai) [reply src/utils/TransactionUtils.ts]: Good call, i implemented with a default retries number we could configure in the future

File: src/utils/TransactionUtils.ts [hunk 2]
```diff
@@ -47,13 +49,31 @@ export async function runTransaction(
...
-      notificationPath: "across-error",
-    });
-    throw error;
+    if (
+      retriesRemaining > 0 &&
+      (error?.code === "NONCE_EXPIRED" ||
+        error?.code === "INSUFFICIENT_FUNDS" ||
+        error?.message.includes("intrinsic gas too low"))
+    ) {
+      // If error is due to a nonce collision or gas underpricement then re-submit to fetch latest params.
```
Comment (mrice32) [inline src/utils/TransactionUtils.ts:58]: Will runTransaction automatically fetch the latest nonce or do we need to manually compute it?

**Review by mrice32 — COMMENTED** [review]
Looks great! Only one question!

File: src/utils/TransactionUtils.ts (hunk 2 above)
Comment (nicholaspai) [reply src/utils/TransactionUtils.ts:58]: it should automatically fetch it as it does now right?

Comment (mrice32) [reply src/utils/TransactionUtils.ts:58]: That's what I'm unsure about. Does ethers grab the nonce at the initial connection time and never requery it? Or does it query it on an interval? Or does it query it on every transaction just before sending? I'll try to look at the provider stack to get a better understanding of how the autopopulation works.

**Review by mrice32 — COMMENTED** [review]
https://github.com/across-protocol/relayer-v2/pull/462#discussion_r1120637185

File: src/utils/TransactionUtils.ts (hunk 2 above)
Comment (nicholaspai) [reply src/utils/TransactionUtils.ts:58]: hm i'm not sure perhaps its better to just manually reset it higher in this case if there is uncertainty

Comment (mrice32) [reply src/utils/TransactionUtils.ts:58]: Okay, I think your assumption is right @nicholaspai. I did some digging in ethers and found [this](https://github.com/ethers-io/ethers.js/blob/f97b92bbb1bde22fcc44100af78d7f31602863ab/packages/abstract-signer/src.ts/index.ts#L293), which implies that the nonce is pulled each time the transaction is sent, and it includes "pending" transactions.

File: src/utils/TransactionUtils.ts [hunk 3]
```diff
@@ -47,13 +49,40 @@ export async function runTransaction(
...
+        retriesRemaining,
+      });
+      return await runTransaction(
+        logger,
+        contract,
+        method,
+        args,
+        value,
+        gasLimit,
+        await contract.signer.getTransactionCount(),
```
Comment (mrice32) [inline src/utils/TransactionUtils.ts:73]: nit: I take back my original point. You can revert this change, and it should work! see https://github.com/across-protocol/relayer-v2/pull/462#discussion_r1120805208

**Review by mrice32 — APPROVED** [review]
LGTM!
